### ISSUE TITLE
`Puppeteer.LaunchAsync` hangs every once in a while waiting for a page to open

### ISSUE THREAD

**Issue by danports** [hardkoded/puppeteer-sharp#2202]
### Description

`Puppeteer.LaunchAsync` hangs occasionally - perhaps once every thousand calls (not sure of the exact frequency).

### Complete minimal example reproducing the issue

I could try writing sample code to launch the browser repeatedly, haven't done that yet.

### Expected behavior:

`Puppeteer.LaunchAsync` should return (or at least throw an exception).

### Actual behavior:

`Puppeteer.LaunchAsync` never returns. It appears from examining the process tree that Chrome launches successfully, but PuppeteerSharp never completes the connection process. Downloads are not a factor here since we're launching a preinstalled binary.

### Versions

* Which version of PuppeteerSharp are you using? 9.1.0
* Which .NET runtime and version are you targeting? 7.0

### Additional Information

I've experienced this issue only in an Ubuntu container, but since I haven't tried running PuppeteerSharp heavily on Windows, I'm not sure whether the OS is a factor.

If I `exec` into the container and kill the Chrome process, Puppeteer throws the following exception:
```
PuppeteerSharp.ProcessException: Failed to create connection  ---> System.TimeoutException: Timeout of 30000 ms exceeded
    at PuppeteerSharp.Helpers.TaskHelper.WithTimeout[T](Task`1 task, TimeSpan timeout, Func`2 exceptionFactory) in /home/runner/work/puppeteer-sharp/puppeteer-sharp/lib/PuppeteerSharp/Helpers/TaskHelper.cs:line 180
    at PuppeteerSharp.Browser.WaitForTargetAsync(Func`2 predicate, WaitForOptions options) in /home/runner/work/puppeteer-sharp/puppeteer-sharp/lib/PuppeteerSharp/Browser.cs:line 235
    at PuppeteerSharp.Launcher.LaunchAsync(LaunchOptions options) in /home/runner/work/puppeteer-sharp/puppeteer-sharp/lib/PuppeteerSharp/Launcher.cs:line 83
    --- End of inner exception stack trace ---
    at PuppeteerSharp.Launcher.LaunchAsync(LaunchOptions options) in /home/runner/work/puppeteer-sharp/puppeteer-sharp/lib/PuppeteerSharp/Launcher.cs:line 89
    at PuppeteerSharp.Launcher.LaunchAsync(LaunchOptions options) in /home/runner/work/puppeteer-sharp/puppeteer-sharp/lib/PuppeteerSharp/Launcher.cs:line 95
```

So it seems that PuppeteerSharp is hanging waiting for a page to open, and the timeout it sets isn't actually being triggered. If it's possible to fix this so that the timeout actually throws, that would be a great start, since it's not a big deal to me if PuppeteerSharp fails to launch every once in a while.

Just reviewing `Browser.WaitForTargetAsync` without knowing much about the PuppeteerSharp internals it seems like there is the potential for a race condition if the `TargetCreated` event is fired just after `Targets()` is called on line 224 and just before the event handler is attached on line 232. However, I'm not sure why the timeout isn't eventually throwing in that case. If that is the issue, it seems like it would be easy to reproduce by inserting a `Task.Delay` between lines 224 and 232.

**Comment by tom2williams** [comment]
I confirm this also happens on Windows systems. Tried with Puppeteer v10.0.0 and different version of Chrome (109, 112, 113). Puppeteer.LaunchAsync hangs (not respecting the timeout set) maybe 3-4 times in 1.000 tries.

**Comment by tom2williams** [comment]
Upon further debugging (I completely removed this next line), it seems that the problem is not:

`await browser.WaitForTargetAsync(t => t.Type == TargetType.Page).ConfigureAwait(false);`

Puppeteer hangs on one of the following lines:


```
                    connection = await Connection
                        .Create(Process.EndPoint, options, _loggerFactory)
                        .ConfigureAwait(false);

                    var browser = await Browser
                        .CreateAsync(
                            options.Product,
                            connection,
                            Array.Empty<string>(),
                            options.IgnoreHTTPSErrors,
                            options.DefaultViewport,
                            Process,
                            null,
                            options.TargetFilter)
                        .ConfigureAwait(false);
```

I will try to debug further.

**Comment by tom2williams** [comment]
After 10k test runs, managed to narrow it down to the following line of code in ChromeTargetManager.cs:

`await _initializeCompletionSource.Task.ConfigureAwait(false);`

(from `public async Task InitializeAsync()` method)

Does anyone know why would this hang and how could we make it timeout somehow?

**Comment by danports** [comment]
Nice work @tom2williams! That task is completed here, so I'm guessing that this line of code is never executed:
https://github.com/hardkoded/puppeteer-sharp/blob/56d0f7c37ec796ebb09b7e3f7982b9a5ea27543e/lib/PuppeteerSharp/ChromeTargetManager.cs#L336

**Comment by leonardo-fernandes** [comment]
@tom2williams could you share how you're reproducing this? I tried calling Puppeteer.LaunchAsync 1000 times, awaiting in between each call, and didn't get any hangs.

**Comment by tom2williams** [comment]
@leonardo-fernandes I am running Puppeteer in a heavily used production environment. There are, at any given time, at least 3-4 instances of Puppeteer already running and calls keep coming. I was not able to reproduce this into a separate app unfortunately. I can add more traces to the code. Just let me know what should I check.

 @danports Indeed, that line of code is not execute, but since the method is executed, it means that _targetsIdsForInit.Count is not 0 (when this hangs). What does that mean?

**Comment by leonardo-fernandes** [comment]
I was able to replicate this fairly consistently with a Main code like the following:

```
Task<IBrowser>[] browserPromises = new Task<IBrowser>[1000];
for (var i = 0; i < browserPromises.Length; i++)
{
    browserPromises[i] = Puppeteer.LaunchAsync(new LaunchOptions() { Product = Product.Chrome, Headless = true }, null);

    if (i >= 5)
    {
        var browser = browserPromises[i - 5].GetAwaiter().GetResult();
        browser.DisposeAsync().GetAwaiter().GetResult();
    }
}

for (var i = 0; i < browserPromises.Length; i++)
{
    var browser = browserPromises[i].GetAwaiter().GetResult();
    browser.DisposeAsync().GetAwaiter().GetResult();
}
```

It launches up to 5 browsers at once. When CPU stays constant at 0%, I pause the debugger and inspect the remaining threads.

I have narrowed it down to a race condition between `_targetsIdsForInit.Remove(target.TargetId);` inside `OnAttachedToTarget` and  `_targetsIdsForInit.Add(kv.Key);` inside `StoreExistingTargetsForInit`. It seems that these two methods run in separate threads, and the following sequencing of steps causes the browser to hang:

* Thread 2 executes `_targetsIdsForInit.Remove(target.TargetId)` - note, the Remove happens before the Add
* Thread 1 executes `_targetsIdsForInit.Add(kv.Key);`
* Thread 2 continues execution and calls `FinishInitializationIfReady();`, however the condition `_targetsIdsForInit.Count == 0` is not satisfied and `_initializeCompletionSource` is never set.
